affinity/cpp-20/d0796r1.md

```diff
@@ -95,395 +96,423 @@ Some systems give additional user control through explicit binding of threads to
-* Defining memory placement algorithms or policies
+* How to migrate memory work and memory allocations between execution resources.
+* How to support dynamic topology discovery and fault tolerance.
 
-# Proposed Wording
+### Querying and representing the system topology
 
-## Header `<execution>` synopsis
+The first task in allowing C++ applications to leverage memory locality is to provide the ability to query a **system** for its **resource topology** (commonly represented as a tree or graph) and traverse its **execution resources**.
 
-    namespace std {
-    namespace experimental {
-    namespace execution {
+The capability of querying underlying **execution resources** of a given **system** is particularly important towards supporting affinity control in C++. The current proposal for executors [5] leaves the **execution resource** largely unspecified. This is intentional: **execution resources** will vary greatly between one implementation and another, and it is out of the scope of the current executors proposal to define those. There is [current work][p0737r0] on extending the executors proposal to describe a typical interface for an **execution context**. In this paper a typical **execution context** is defined with an interface for construction and comparison, and for retrieving an **executor**, waiting on submitted work to complete and querying the underlying **execution resource**. Extending the executors interface to provide topology information can serve as a basis for providing a unified interface to expose affinity. This interface cannot mandate a specific architectural definition, and must be generic enough that future architectural evolutions can still be expressed.
 
-    /* Execution resource */
+Two important considerations when defining a unified interface for querying the **resource topology** of a **system**, are (a) what level of abstraction such an interface should have, and (b) at what granularity it should describe the topology's **execution resources**. As both the level of abstraction of an **execution resource** and the granularity that it is described in will vary greatly from one implementation to another, it’s important for the interface to be generic enough to support any level of abstraction. To achieve this we propose a generic hierarchical structure of **execution resources**, each **execution resource** being composed of other **execution resources** recursively. Each **execution resource** within this hierarchy can be used to place memory (i.e., allocate memory within the **execution resource’s** memory region), place execution (i.e. bind an execution to an **execution resource’s execution agents**), or both.
 
-    struct execution_resource {
+> For example, a NUMA system will likely have a hierarchy of nodes, each capable of placing memory and placing agents.  A CPU + GPU system may have GPU local memory regions capable of placing memory, but not capable of placing agents.
```

Comment:
there is >

Reply:
This was indentation for the example, but we don't really use this format anywhere else so I'll just remove it.